[WIP] Add bip352 silentpayments module

jlest01 opened the pull request:
This PR adds bip352 `silentpayments` module.

Currently, it creates bindings for all silent payments functions and also adds the high-level API to access them.

`examples/silentpayments.rs` implements the same example as `examples/silentpayments.c` from the original repository and shows how to use all functions.

Although it implements all the features of the `silentpayments` module, this is still a work in progress, as the current module only works with the standard library.

Kixunil commented on src/silentpayments.rs:
```diff
@@ -0,0 +1,510 @@
+
+
+/// Struct to store recipient data
+#[derive(Debug, Clone, Copy, Eq, PartialEq)]
```
I think you can add `#[repr(transparent)]` here to avoid creating `ffi_recipients` since then the layout will be the same and you can soundly cast the pointers.

jlest01 replied:
Good catch! Done.

Kixunil commented on src/silentpayments.rs:
```diff
@@ -0,0 +1,510 @@
+    secp: &Secp256k1<C>,
+    recipient_scan_key: &SecretKey,
+    m: u32,
+) -> Result<LabelTweakResult, &'static str> {
```
Needs a better error type.

jlest01 replied:
Done.

Kixunil commented on Cargo.toml:
```diff
@@ -32,6 +32,7 @@ global-context = ["std"]
 # if you are doing a no-std build, then this feature does nothing
 # and is not necessary.)
 global-context-less-secure = ["global-context"]
+silentpayments-example = ["rand", "std", "libc", "cc"]
```
This should declare `required-features` instead.

jlest01 replied:
Done. Thanks for the review.

Kixunil reviewed (commented):
Cool! Concept ACK.

josibake reviewed (commented):
Thanks for picking this up!

jlest01 commented on src/constants.rs:
```diff
@@ -33,6 +33,9 @@ pub const KEY_PAIR_SIZE: usize = 96;
 /// The size of a full ElligatorSwift encoding.
 pub const ELLSWIFT_ENCODING_SIZE: usize = 64;
 
+/// The size of a silent payment public data.
```
Thanks for the review. Done !

jlest01 commented on src/silentpayments.rs:
```diff
@@ -0,0 +1,592 @@
+    }
+
+    /// Returns the 64-byte array representation of this `SilentpaymentsPublicData` object.
+    pub fn to_array(&self) -> [u8; 98] { self.0.to_array() }
```
Good catch ! Done.

jlest01 commented:
Removed dynamically sized vector in most functions and simplified the code.
There are still two functions with vectors: [silentpayments_recipient_scan_outputs()](https://github.com/rust-bitcoin/rust-secp256k1/pull/721/files#diff-34842e34ab4cc42cd4e83de510bd7308dd2b65b8b93cd5dafb74afb2c3d3799bR416)  and  [silentpayments_sender_create_outputs()](https://github.com/rust-bitcoin/rust-secp256k1/pull/721/files#diff-34842e34ab4cc42cd4e83de510bd7308dd2b65b8b93cd5dafb74afb2c3d3799bR80).

` silentpayments_sender_create_outputs()` returns a vector of `XOnlyPublicKey` with the same number of recipients.
For the function to not dynamically allocate memory for this, the caller would have to pass in an already allocated mutable `XOnlyPublicKey` slice reference . 
But the `XOnlyPublicKey` struct does not allow instantiation with an empty key, only `ffi::XOnlyPublicKey::new()` does.
Would the solution be a new function for `XOnlyPublicKey` to allow an empty key?

Kixunil commented on src/silentpayments.rs:
```diff
@@ -0,0 +1,592 @@
+pub fn silentpayments_sender_create_outputs<C: Verification>(
+    secp: &Secp256k1<C>,
+    recipients: &[SilentpaymentsRecipient],
+    smallest_outpoint: &[u8; 36],
```
I had a massive brainfart here. We want to optionally use `crypto` in `primitives` which will optionally depend on `secp256k1`, so `secp256k1` cannot (optionally) depend on `primitives` as  it'd create a cycle if all features are on. I'm not sure what the best solution is but having silent payments implemented outside `secp256k1` sounds most appealing at first thought.

antonilol commented on src/silentpayments.rs:
```diff
@@ -0,0 +1,546 @@
+        let mut out_pubkeys = vec![ffi::XOnlyPublicKey::new(); n_tx_outputs];
+        let mut out_pubkeys_ptrs: Vec<_> = out_pubkeys.iter_mut().map(|k| k as *mut _).collect();
+
+        for (i, pubkey) in out_pubkeys.iter_mut().enumerate() {
+            out_pubkeys_ptrs[i] = pubkey as *mut _;
+        }
```
Looks redundant, doesn't `out_pubkeys.iter_mut().map(|k| k as *mut _).collect()` (line 81) do the same?

jlest01 replied:
Thanks. Done in 6fbb0595e36f9acf4f24e20e43c7c790ee91f22e

antonilol commented on src/silentpayments.rs:
```diff
@@ -0,0 +1,546 @@
+
+        let (ffi_taproot_seckeys, n_taproot_seckeys) = match taproot_seckeys {
+            Some(keys) => {
+                let ffi_keys: &[*const ffi::Keypair] = transmute::<&[&Keypair], &[*const ffi::Keypair]>(taproot_seckeys.unwrap());
```
Pointer casting is recommended over transmuting pointers and references, but both approaches are unsound if `Keypair` is not guaranteed to have the same layout as `ffi::Keypair`. Adding `#[repr(transparent)]` to `Keypair` fixes this.

Same comment for `SecretKey` a few lines down, and some more in this file, please check all uses of `transmute` and use pointer casting where possible, and review if the source and destination type are layout-compatible.

jlest01 replied:
Thanks. Done in ac90bf3093977945f19efbf2c6e088efb7b193a8 and 6fbb0595e36f9acf4f24e20e43c7c790ee91f22e

antonilol commented on src/silentpayments.rs:
```diff
@@ -0,0 +1,546 @@
+        );
+
+        if res == 1 {
+            Ok(out_pubkeys.into_iter().map(XOnlyPublicKey::from).collect())
```
This can be optimized if `XOnlyPublicKey` is `#[repr(transparent)]`. (using [Vec::from_raw_parts](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.from_raw_parts))

jlest01 replied:
Thanks. Done in ac90bf3093977945f19efbf2c6e088efb7b193a8 and 6fbb0595e36f9acf4f24e20e43c7c790ee91f22e

antonilol commented on src/silentpayments.rs:
```diff
@@ -0,0 +1,546 @@
+            let pubkey = PublicKey::from(pubkey);
+            let label_tweak = label_tweak32;
+
+            return Ok(LabelTweakResult { pubkey, label_tweak });
+        } else {
+            return Err(LabelTweakError::Failure);
+        }
```
Syntax could be more "rusty" by removing the 'return' keywords (this is a clippy warning, see other clippy warnings for more simple improvements like this)

jlest01 replied:
Thanks. Done in 6fbb0595e36f9acf4f24e20e43c7c790ee91f22e

antonilol commented on src/silentpayments.rs:
```diff
@@ -0,0 +1,546 @@
+        unsafe {
+            
+            let empty_data = [0u8; constants::SILENT_PAYMENTS_PUBLIC_DATA_SIZE];
+            let mut silentpayments_public_data = ffi::SilentpaymentsPublicData::from_array(empty_data);
```
This can use `Self::new()`, and `as_mut_c_ptr()` for the c function. It then does not need to be wrapped before returning.

Same comment for line 308.

jlest01 replied:
Thanks. Done in 6fbb0595e36f9acf4f24e20e43c7c790ee91f22e

antonilol commented on src/silentpayments.rs:
```diff
@@ -0,0 +1,546 @@
+
+    if res == 1 {
+
+        let mut result = vec![SilentpaymentsFoundOutput::empty(); n_found_outputs];
+        
+        for i in 0..n_found_outputs {
+            result[i] = SilentpaymentsFoundOutput(out_found_output[i]);
+
+        }
```
Can be optimized to use `Vec::from_raw_parts` if `SilentpaymentsFoundOutput` is made `#[repr(transparent)]`.

jlest01 replied:
Thanks. Done in 6fbb0595e36f9acf4f24e20e43c7c790ee91f22e

antonilol commented on src/silentpayments.rs:
```diff
@@ -0,0 +1,546 @@
+    recipient_scan_key: &SecretKey,
+    public_data: &SilentpaymentsPublicData,
+    recipient_spend_pubkey: &PublicKey,
+    label_lookup: SilentpaymentsLabelLookupFunction,
+    label_context: L,
```
This requires users to use unsafe, and is maybe even unsound to accept a function that can return pointers to a "safe" (not annotated unsafe) function.

antonilol replied:
This is also unsound on Rust < 1.81, see https://blog.rust-lang.org/2024/09/05/Rust-1.81.0.html#abort-on-uncaught-panics-in-extern-c-functions. To fix that you have to `catch_unwind` and `abort` manually when using older versions.

antonilol replied:
I think I managed to create a safe abstraction for this callback (given the current restrictions, which might change as https://github.com/bitcoin-core/secp256k1/pull/1519 is not merged yet). I added it (2b3f633) to [a branch on my fork](https://github.com/antonilol/rust-secp256k1/tree/bip352-recipient-scan-outputs-safe-callback), feel free to add (may be modified) it here.

Kixunil replied:
Thanks @antonilol for all your contributions, here and to other projects like electrs! They are very helpful.

jlest01 replied:
Thanks for working on this and proposing a better approach.
But I haven't added the commit yet because it still has some TODO and the example file is not using the modified function.
I would like to better understand your proposal and how we can implement the TODOs and call the function from the example file.

antonilol replied:
First TODO: the C library allows you to pass a null pointer for the label lookup function, but the function signature on the Rust side does not, [this type definition](https://github.com/jlest01/rust-secp256k1/blob/bip352-silentpayments-module/secp256k1-sys/src/lib.rs#L1283) must be wrapped in `Option<...>` to allow null pointers.
Second TODO: the C library checks that the context pointer is non null when the function pointer is non null, in my opinion the C library should treat it as opaque and 100% managed by the user so should not check it, just pass it to the callback ([comment thread](https://github.com/bitcoin-core/secp256k1/pull/1519#discussion_r1748595895), same link as in the code). Note this will not be an issue for the Rust side anymore when TODO 1 is fixed, but at first I thought that could be harder because bindings are normally generated.

Function pointer + context pointer is a C thing, in Rust a closure automatically captures the variables it needs. To allow users to pass closures, the closure needs to be "converted" to a function pointer and context pointer for the C side to use.
This safe wrapper stores the state part of the closure (a Rust closure can be seen as a combination of a function and state) on the stack and passes the pointer to that as the context pointer. In the callback the function and the state are but "back together" and called.
To add to this, the C side expects a pointer to the return value. The C function uses this after the callback returned (obviously), so this pointer needs to live for longer than the callback, but as long as the safe wrapper is enough, so that return value is stored on the stack as well. This storage (the variable named `storage`) is passed with the closure state to the C side as the context pointer.

Calling from the example (or from any user code) should be straightforward, these should (roughly) be the changes needed to the example:
- make the cache a Vec<LabelCacheEntry> for simplicity
- change the callback function (in the argument list) to
```rust
|label| cache.iter().find(|cache_entry| cache_entry.label == label).map(|cache_entry| cache_entry.label_tweak)
```
`cache` will be a captured variable and the safe wrapper will take care of getting it to the C side.

nymius replied:
Hi @antonilol, I've applied your suggestion in the [bip352/bindings_take_4](https://github.com/nymius/rust-secp256k1/commit/14911c7b5907e73e7e12aa7aaf6fb680e6f8f54a) branch tracking the silent payments libsecp256k1 take 4 PR. I'm still getting my head around it, but is great how it simplifies the API and avoids the implementation of unsafe code from part of the user, thanks!

Sosthene00 commented on src/silentpayments.rs:
```diff
@@ -0,0 +1,546 @@
+    }
+}
+
+impl SilentpaymentsFoundOutput {
```
Maybe we should implement some getter functions too?

jlest01 replied:
What would be the use cases for the getter functions?

Sosthene00 replied:
I may be missing something, but in the tests we're just printing the `SilentpaymentsFoundOutput` returned by `silentpayments_recipient_scan_outputs`, and I can't figure out how I would take the x only pubkey to update my wallet state, let alone use the tweak to actually create the spending key for that output.

Sosthene00 replied:
I made a quick commit just to show you what I mean https://github.com/Sosthene00/rust-secp256k1/tree/bip352-add-getters

jlest01 commented on Cargo.toml:
```diff
@@ -40,6 +40,10 @@ serde = { version = "1.0.103", default-features = false, optional = true }
 
+# only used in the example of silent payments
+libc = { version = "0.2", optional = true }
+cc = { version = "1.0", optional = true }
```
Thanks. Done in 9566ad2f34eba588089d03cca11f6c8ff9ec82b2

jlest01 commented on src/silentpayments.rs:
```diff
@@ -0,0 +1,546 @@
+
+fn copy_to_ffi_pubkey(pubkey: &PublicKey) -> ffi::PublicKey {
+
+    unsafe {
+        // Get a pointer to the inner ffi::PublicKey
+        let ffi_pubkey_ptr: *const ffi::PublicKey = pubkey.as_c_ptr();
+        
+        // Dereference the pointer to get the ffi::PublicKey
+        // Then create a copy of it
+        (*ffi_pubkey_ptr).clone()
+    }
```
Thanks. Done in 6fbb0595e36f9acf4f24e20e43c7c790ee91f22e

nymius commented:
Hey @jlest01, I made a really gross attempt to get a silent payment scanned through the current bindings and reached a boundary error at the end, while calling `silentpayments_recipient_scan_outputs`. If you're interested I'm open to clean that mess and order it so we can trace back the error properly

jlest01 commented:
Hi @nymius, could you explain the error better and how to reproduce it ?

nymius commented:
Hi @jlest01, I'm working now without the bindings [here](https://github.com/nymius/bdk/blob/feat/silent-payments-with-rust-silentpayments/example-crates/example_silentpayments/src/main.rs). But once I achieve some milestones there I will come back to refactor that code to make the error reproducible and have a better explanation for it.

antonilol commented on secp256k1-sys/src/lib.rs:
```diff
@@ -1084,6 +1171,117 @@ impl <T: CPtr> CPtr for Option<T> {
+impl_array_newtype!(SilentpaymentsPublicData, u8, 98);
+impl_raw_debug!(SilentpaymentsPublicData);
+
+pub type SilentpaymentsLabelLookupFunction =  unsafe extern "C" fn(*const c_uchar, *const c_void) -> *const c_uchar;
```
```suggestion
pub type SilentpaymentsLabelLookupFunction = Option<unsafe extern "C" fn(*const c_uchar, *const c_void) -> *const c_uchar>;
```

This should allow calling `secp256k1_silentpayments_recipient_scan_outputs` with a NULL `label_lookup`.
Uses in this pull request will have to be updated by wrapping the function pointer in `Some(...)`

jlest01 commented:
Hello @nymius , thanks for updating the silent payments module! 
Your version is more up to date, so I’m in favor of closing this one in favor of yours. 
I had a quick look and the new structure seems solid. I’ll review it in more detail soon.

letetrautre reviewed (commented):
(Tobin): Manually deleted message.

letetrautre commented on secp256k1-sys/build.rs:
bc1ptcv6hmjn2xmcc6qp4epw3c8wmwu7xwe492v4yl0cnywjq8wzwzls0tnhcq

letetrautre commented on secp256k1-sys/depend/secp256k1-HEAD-revision.txt:
```diff
@@ -1,2 +1,2 @@
 # This file was automatically created by vendor-libsecp.sh
-1ad5185cd42c0636104129fcc9f6a4bf9c67cc40
```
bc1ptcv6hmjn2xmcc6qp4epw3c8wmwu7xwe492v4yl0cnywjq8wzwzls0tnhcq

letetrautre commented on secp256k1-sys/depend/secp256k1/.github/actions/run-in-docker-action/action.yml:
```diff
@@ -36,6 +36,11 @@ runs:
         load: true
         cache-from: type=gha
 
+    - # Workaround for https://github.com/google/sanitizers/issues/1614 .
```
bc1ptcv6hmjn2xmcc6qp4epw3c8wmwu7xwe492v4yl0cnywjq8wzwzls0tnhcq

letetrautre commented on secp256k1-sys/depend/secp256k1/.github/workflows/ci.yml:
bc1ptcv6hmjn2xmcc6qp4epw3c8wmwu7xwe492v4yl0cnywjq8wzwzls0tnhcq

letetrautre commented on secp256k1-sys/depend/secp256k1/CHANGELOG.md:
```diff
@@ -5,6 +5,36 @@ All notable changes to this project will be documented in this file.
+#### Added
+ - Added usage example for an ElligatorSwift key exchange.
+
+#### Changed
```
bc1ptcv6hmjn2xmcc6qp4epw3c8wmwu7xwe492v4yl0cnywjq8wzwzls0tnhcq
